// @skyway-sdk/room/dist/room/base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomBase = void 0;
const common_1 = require("@skyway-sdk/common");
const sfu_bot_1 = require("@skyway-sdk/sfu-bot");
const errors_1 = require("../errors");
const base_1 = require("../member/remote/base");
const publication_1 = require("../publication");
const subscription_1 = require("../subscription");
const util_1 = require("../util");
const log = new common_1.Logger('packages/room/src/room/base.ts');
/**@internal */
class RoomBase {
    constructor(type, _channel) {
        this._channel = _channel;
        this._members = {};
        this._publications = {};
        this._subscriptions = {};
        this._context = this._channel._context;
        this._events = new common_1.Events();
        this.onClosed = this._events.make();
        this.onMetadataUpdated = this._events.make();
        this.onMemberJoined = this._events.make();
        this.onMemberLeft = this._events.make();
        this.onMemberListChanged = this._events.make();
        this.onMemberMetadataUpdated = this._events.make();
        this.onStreamPublished = this._events.make();
        this.onStreamUnpublished = this._events.make();
        this.onPublicationListChanged = this._events.make();
        this.onPublicationMetadataUpdated = this._events.make();
        this.onPublicationEnabled = this._events.make();
        this.onPublicationDisabled = this._events.make();
        this.onPublicationSubscribed = this._events.make();
        this.onPublicationUnsubscribed = this._events.make();
        this.onSubscriptionListChanged = this._events.make();
        this.type = type;
        this._channel.onClosed.pipe(this.onClosed);
        this._channel.onMetadataUpdated.pipe(this.onMetadataUpdated);
        this._channel.onMemberMetadataUpdated.add((e) => this._handleOnMemberMetadataUpdate(e));
        this._setChannelState();
        this._setChannelListener();
    }
    /**@private */
    static _createBot(context, channel) {
        return __awaiter(this, void 0, void 0, function* () {
            const plugin = context.plugins.find((p) => p.subtype === 'sfu');
            const bot = channel.members.find((m) => m.subtype === sfu_bot_1.SFUBotMember.subtype);
            if (!bot) {
                yield plugin.createBot(channel);
            }
            return plugin;
        });
    }
    /**@private */
    _getMember(id) {
        return this._members[id];
    }
    /**@private */
    _getPublication(id) {
        return this._publications[id];
    }
    /**@private */
    _getOriginPublication(publicationId) {
        const origin = this.publications.find((p) => { var _a; return ((_a = p._publication.origin) === null || _a === void 0 ? void 0 : _a.id) === publicationId; });
        return origin;
    }
    /**@private */
    _addPublication(p) {
        const exist = this._publications[p.id];
        if (exist) {
            return exist;
        }
        const publication = new publication_1.RoomPublicationImpl(p, this);
        this._publications[p.id] = publication;
        return publication;
    }
    /**@private */
    _getSubscription(id) {
        return this._subscriptions[id];
    }
    /**@private */
    _addSubscription(s) {
        const exist = this._subscriptions[s.id];
        if (exist) {
            return exist;
        }
        const subscription = new subscription_1.RoomSubscriptionImpl(s, this);
        this._subscriptions[s.id] = subscription;
        return subscription;
    }
    get id() {
        return this._channel.id;
    }
    get name() {
        return this._channel.name;
    }
    get metadata() {
        return this._channel.metadata;
    }
    get state() {
        return this._channel.state;
    }
    get disposed() {
        return this._channel.disposed;
    }
    _handleOnMemberMetadataUpdate(e) {
        const member = this._getMember(e.member.id);
        this.onMemberMetadataUpdated.emit({ member, metadata: e.metadata });
    }
    _setChannelState() {
        this._channel.members.forEach((m) => {
            if (m.type === 'bot') {
                return;
            }
            const member = new base_1.RemoteRoomMemberImpl(m, this);
            this._members[m.id] = member;
        });
        this._channel.publications.forEach((p) => {
            if (!this._isAcceptablePublication(p)) {
                return;
            }
            this._addPublication(p);
        });
        this._channel.subscriptions.forEach((s) => {
            if (s.subscriber.type === 'bot') {
                return;
            }
            this._addSubscription(s);
        });
    }
    _setChannelListener() {
        this._channel.onMemberJoined.add((e) => this._handleOnMemberJoin(e.member));
        this._channel.onMemberLeft.add((e) => this._handleOnMemberLeft(e.member));
        this._channel.onStreamPublished.add((e) => this._handleOnStreamPublish(e.publication));
        this._channel.onStreamUnpublished.add((e) => this._handleOnStreamUnpublish(e.publication));
        this._channel.onPublicationMetadataUpdated.add((e) => this._handleOnPublicationMetadataUpdate(e.publication));
        this._channel.onPublicationEnabled.add((e) => this._handleOnPublicationEnabled(e.publication));
        this._channel.onPublicationDisabled.add((e) => this._handleOnPublicationDisabled(e.publication));
        this._channel.onPublicationSubscribed.add((e) => this._handleOnStreamSubscribe(e.subscription));
        this._channel.onPublicationUnsubscribed.add((e) => this._handleOnStreamUnsubscribe(e.subscription));
    }
    _handleOnMemberJoin(m) {
        if (m.type === 'bot') {
            return;
        }
        if (this._getMember(m.id)) {
            return;
        }
        const member = new base_1.RemoteRoomMemberImpl(m, this);
        this._members[m.id] = member;
        this.onMemberJoined.emit({ member });
        this.onMemberListChanged.emit({});
    }
    _handleOnMemberLeft(m) {
        const member = this._getMember(m.id);
        if (!member) {
            // should be bot
            return;
        }
        delete this._members[m.id];
        if (m.side === 'remote') {
            member._dispose();
        }
        this.onMemberLeft.emit({ member });
        this.onMemberListChanged.emit({});
    }
    _handleOnStreamPublish(p) {
        if (!this._isAcceptablePublication(p)) {
            return;
        }
        if (this._getPublication(p.id)) {
            return;
        }
        const publication = this._addPublication(p);
        this.onStreamPublished.emit({ publication });
        this.onPublicationListChanged.emit({});
    }
    _handleOnStreamUnpublish(p) {
        if (!this._isAcceptablePublication(p)) {
            return;
        }
        const publication = this._getPublication(p.id);
        delete this._publications[p.id];
        this.onStreamUnpublished.emit({ publication });
        this.onPublicationListChanged.emit({});
    }
    _handleOnPublicationMetadataUpdate(p) {
        const publication = this._getTargetPublication(p.id, p.type);
        if (!publication)
            return;
        this.onPublicationMetadataUpdated.emit({
            publication,
            metadata: publication.metadata,
        });
    }
    _handleOnPublicationEnabled(p) {
        const publication = this._getTargetPublication(p.id, p.type);
        if (!publication)
            return;
        this.onPublicationEnabled.emit({ publication });
    }
    _handleOnPublicationDisabled(p) {
        const publication = this._getTargetPublication(p.id, p.type);
        if (!publication)
            return;
        this.onPublicationDisabled.emit({ publication });
    }
    _handleOnStreamSubscribe(s) {
        if (s.subscriber.type === 'bot') {
            return;
        }
        if (this._getSubscription(s.id)) {
            return;
        }
        const subscription = this._addSubscription(s);
        this.onPublicationSubscribed.emit({ subscription });
        this.onSubscriptionListChanged.emit({});
    }
    _handleOnStreamUnsubscribe(s) {
        if (s.subscriber.type === 'bot') {
            return;
        }
        const subscription = this._getSubscription(s.id);
        delete this._subscriptions[s.id];
        this.onPublicationUnsubscribed.emit({ subscription });
        this.onSubscriptionListChanged.emit({});
    }
    get members() {
        return Object.values(this._members);
    }
    get publications() {
        return Object.values(this._publications);
    }
    get subscriptions() {
        return Object.values(this._subscriptions);
    }
    _joinChannel(roomMemberInit = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.state !== 'opened') {
                throw (0, util_1.createError)({
                    operationName: 'RoomImpl.joinChannel',
                    context: this._context,
                    room: this,
                    info: errors_1.errors.roomNotOpened,
                    path: log.prefix,
                });
            }
            const local = yield this._channel.join(roomMemberInit);
            if (!this._getMember(local.id)) {
                yield this.onMemberJoined
                    .watch((e) => {
                    return e.member._member.id === local.id;
                }, this._context.config.rtcApi.timeout)
                    .catch((error) => {
                    throw (0, util_1.createError)({
                        operationName: 'RoomImpl.joinChannel',
                        context: this._context,
                        room: this,
                        info: Object.assign(Object.assign({}, errors_1.errors.timeout), { detail: 'RoomImpl onMemberJoined' }),
                        path: log.prefix,
                        error,
                    });
                });
            }
            return local;
        });
    }
    join(memberInit = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const local = yield this._joinChannel(Object.assign(Object.assign({}, memberInit), { disableSignaling: this._disableSignaling }));
            const localRoomMember = this._createLocalRoomMember(local, this);
            log.debug('member joined', memberInit);
            this.localRoomMember = localRoomMember;
            this._members[localRoomMember.id] = localRoomMember;
            localRoomMember.onLeft.once(() => {
                this.localRoomMember = undefined;
            });
            return localRoomMember;
        });
    }
    leave(member) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._channel.leave(member._member);
        });
    }
    updateMetadata(metadata) {
        return this._channel.updateMetadata(metadata);
    }
    close() {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._channel.close();
        });
    }
    dispose() {
        return __awaiter(this, void 0, void 0, function* () {
            return this._channel.dispose();
        });
    }
    toJSON() {
        return {
            type: this.type,
            id: this.id,
            name: this.name,
            metadata: this.metadata,
            members: this.members,
            publications: this.publications,
            subscriptions: this.subscriptions,
        };
    }
}
exports.RoomBase = RoomBase;
