// @skyway-sdk/room/dist/room/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.roomTypes = exports.SkyWayRoom = void 0;
const common_1 = require("@skyway-sdk/common");
const core_1 = require("@skyway-sdk/core");
const sfu_bot_1 = require("@skyway-sdk/sfu-bot");
const errors_1 = require("../errors");
const version_1 = require("../version");
const default_1 = require("./default");
const p2p_1 = require("./p2p");
const sfu_1 = require("./sfu");
const log = new common_1.Logger('packages/room/src/room/index.ts');
class SkyWayRoom {
    /**@private */
    // biome-ignore lint/complexity/noUselessConstructor: Private constructor prevents instantiation of static-only class
    constructor() { }
}
exports.SkyWayRoom = SkyWayRoom;
_a = SkyWayRoom;
/**
 * @description [japanese] Roomの作成
 * RoomInit.typeに応じてRoom, P2PRoom, SFURoomのいずれかとして作成する
 * - 'default'または未指定: Roomとして作成する
 * - 'p2p': P2PRoomとして作成する
 * - 'sfu': SFURoomとして作成する
 */
SkyWayRoom.Create = (context, init) => __awaiter(void 0, void 0, void 0, function* () {
    log.info('room created', {
        operationName: 'SkyWayRoom._Factory',
        sdkName: 'room',
        sdkVersion: version_1.PACKAGE_VERSION,
        init,
    });
    const plugin = new sfu_bot_1.SFUBotPlugin(init === null || init === void 0 ? void 0 : init.sfuOptions);
    context.registerPlugin(plugin);
    const channel = yield core_1.SkyWayChannel.Create(context, {
        name: init.name,
        metadata: init.metadata,
    });
    const room = yield SkyWayRoom._Factory(context, channel, init.type);
    return SkyWayRoom._castRoomType(room);
});
/**
 * @description [japanese] 既存のRoomの取得
 * FindOptions.typeに応じてRoom, P2PRoom, SFURoomのいずれかとして取得する
 * - 'default'または未指定: Roomとして取得する
 * - 'p2p': P2PRoomとして取得する
 * - 'sfu': SFURoomとして取得する
 */
SkyWayRoom.Find = (context, query, options) => __awaiter(void 0, void 0, void 0, function* () {
    const plugin = new sfu_bot_1.SFUBotPlugin(options === null || options === void 0 ? void 0 : options.sfuOptions);
    context.registerPlugin(plugin);
    const channel = yield core_1.SkyWayChannel.Find(context, query);
    const roomType = options === null || options === void 0 ? void 0 : options.type;
    const room = yield SkyWayRoom._Factory(context, channel, roomType);
    return SkyWayRoom._castRoomType(room);
});
/**
 * @description [japanese] Roomの取得を試み、存在しなければ作成する
 * RoomInit.typeに応じてRoom, P2PRoom, SFURoomのいずれかとして作成または取得する
 * - 'default'または未指定: Roomとして作成または取得する
 * - 'p2p': P2PRoomとして作成または取得する
 * - 'sfu': SFURoomとして作成または取得する
 */
SkyWayRoom.FindOrCreate = (context, init) => __awaiter(void 0, void 0, void 0, function* () {
    const plugin = new sfu_bot_1.SFUBotPlugin(init === null || init === void 0 ? void 0 : init.sfuOptions);
    context.registerPlugin(plugin);
    const channel = yield core_1.SkyWayChannel.FindOrCreate(context, Object.assign({}, init));
    const room = yield SkyWayRoom._Factory(context, channel, init.type);
    return SkyWayRoom._castRoomType(room);
});
SkyWayRoom._Factory = (context, channel, roomType) => __awaiter(void 0, void 0, void 0, function* () {
    const type = roomType !== null && roomType !== void 0 ? roomType : 'default';
    switch (type) {
        case 'p2p':
            return new p2p_1.P2PRoomImpl(channel);
        case 'sfu':
            return (yield sfu_1.SFURoomImpl.Create(context, channel));
        case 'default':
            return (yield default_1.RoomImpl.Create(context, channel));
        default:
            throw (0, core_1.createError)({
                operationName: 'SkyWayRoom._Factory',
                context,
                channel,
                info: errors_1.errors.notImplemented,
                path: log.prefix,
            });
    }
});
SkyWayRoom._castRoomType = (room) => {
    return room;
};
exports.roomTypes = ['sfu', 'p2p', 'default'];
