// @skyway-sdk/room/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.errors = exports.roomErrors = void 0;
const core_1 = require("@skyway-sdk/core");
const sfu_bot_1 = require("@skyway-sdk/sfu-bot");
exports.roomErrors = {
    invalidParameter: { name: 'invalidParameter', detail: '', solution: '' },
    timeout: { name: 'timeout', detail: '', solution: '' },
    internal: { name: 'internal', detail: '', solution: '' },
    notImplemented: {
        name: 'notImplemented',
        detail: '対応していないRoomTypeです',
        solution: '正しいRoomTypeを指定してください',
    },
    roomNotOpened: {
        name: 'roomNotOpened',
        detail: 'RoomがOpenされていません',
        solution: 'Roomの状態を確かめてください',
    },
    subscribeOtherMemberType: {
        name: 'subscribeOtherMemberType',
        detail: 'RemoteMemberにSubscribe/Unsubscribeさせる場合、対象のMemberのTypeはPersonである必要があります',
        solution: '対象のRemoteMemberが正しいか確認してください',
    },
    sfuPublicationNotSupportDataStream: {
        name: 'sfuPublicationNotSupportDataStream',
        detail: 'RoomでPublicationのtypeを"sfu"に設定したり、SFURoomを使用したりする場合、DataStreamを使うことは出来ません',
        solution: 'Roomを使用して、DataStreamのpublishのoptions.typeに"p2p"を指定してください',
    },
    publicationNotHasOrigin: {
        name: 'publicationNotHasOrigin',
        detail: 'SFURoomで操作するPublicationはOriginをもつ必要があります',
        solution: 'SFURoomとP2PRoomを同一のIDで混在させていないか確かめてください',
    },
    notFound: {
        name: 'notFound',
        detail: '参照しようとしていたものが見つかりません',
        solution: '参照しようとしたものが存在するか確かめてください',
    },
    invalidPublicationTypeForP2PRoom: {
        name: 'invalidPublicationTypeForP2PRoom',
        detail: 'P2PRoomでSFUのPublicationをpublishすることは出来ません',
        solution: 'publishのoptions.typeに"p2p"を指定してください',
    },
    invalidPublicationTypeForSFURoom: {
        name: 'invalidPublicationTypeForSFURoom',
        detail: 'SFURoomでP2PのPublicationをpublishすることは出来ません',
        solution: 'publishのoptions.typeに"sfu"を指定してください',
    },
};
exports.errors = Object.assign(Object.assign(Object.assign({}, core_1.errors), sfu_bot_1.errors), exports.roomErrors);
