// @skyway-sdk/room/dist/subscription/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoomSubscriptionImpl = void 0;
const common_1 = require("@skyway-sdk/common");
const core_1 = require("@skyway-sdk/core");
const log = new common_1.Logger('packages/room/src/subscription/index.ts');
/**@internal */
class RoomSubscriptionImpl {
    constructor(
    /**@private */
    _subscription, _room) {
        this._subscription = _subscription;
        this._room = _room;
        this._context = this._room._context;
        this.onStreamAttached = new core_1.Event();
        this.onConnectionStateChanged = new core_1.Event();
        this.id = _subscription.id;
        this.contentType = _subscription.contentType;
        this.publication = this._room._getPublication(_subscription.publication.id);
        this.subscriber = this._room._getMember(_subscription.subscriber.id);
        _subscription.onStreamAttached.pipe(this.onStreamAttached);
        _subscription.onConnectionStateChanged.add((state) => {
            log.debug('_subscription.onConnectionStateChanged', this.id, state);
            this.onConnectionStateChanged.emit(state);
        });
    }
    get stream() {
        return this._subscription.stream;
    }
    get state() {
        return this._subscription.state;
    }
    get codec() {
        return this._subscription.codec;
    }
    get preferredEncoding() {
        return this._subscription.preferredEncoding;
    }
    changePreferredEncoding(id) {
        this._subscription.changePreferredEncoding(id);
    }
    toJSON() {
        return {
            id: this.id,
            contentType: this.contentType,
            publication: this.publication,
            codec: this.codec,
        };
    }
    getStats() {
        return this._subscription.getStats();
    }
    getRTCPeerConnection() {
        return this._subscription.getRTCPeerConnection();
    }
    getConnectionState() {
        return this._subscription.getConnectionState();
    }
}
exports.RoomSubscriptionImpl = RoomSubscriptionImpl;
