// @skyway-sdk/room/dist/member/local/base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalRoomMemberBase = void 0;
const common_1 = require("@skyway-sdk/common");
const core_1 = require("@skyway-sdk/core");
const sfu_bot_1 = require("@skyway-sdk/sfu-bot");
const const_1 = require("../../const");
const errors_1 = require("../../errors");
const member_1 = require("../../member");
const log = new common_1.Logger('packages/room/src/member/local/base.ts');
/**@internal */
class LocalRoomMemberBase extends member_1.RoomMemberImpl {
    /**@private */
    constructor(member, room) {
        super(member, room);
        this.side = 'local';
        this._local = this._member;
        this.onStreamPublished = new common_1.Event();
        this.onStreamUnpublished = new common_1.Event();
        this.onPublicationListChanged = new common_1.Event();
        this.onPublicationSubscribed = new common_1.Event();
        this.onPublicationUnsubscribed = new common_1.Event();
        this.onSubscriptionListChanged = new common_1.Event();
        this.onFatalError = new common_1.Event();
        this._context = this.room._context;
        this._local.onPublicationSubscribed.add((e) => __awaiter(this, void 0, void 0, function* () {
            const roomSubscription = room._addSubscription(e.subscription);
            this.onPublicationSubscribed.emit({
                subscription: roomSubscription,
                stream: e.stream,
            });
        }));
        this._local.onFatalError.pipe(this.onFatalError);
        this._listenRoomEvent();
        this.onStreamPublished.add(() => this.onPublicationListChanged.emit());
        this.onStreamUnpublished.add(() => this.onPublicationListChanged.emit());
        this.onPublicationSubscribed.add(() => this.onSubscriptionListChanged.emit());
        this.onPublicationUnsubscribed.add(() => this.onSubscriptionListChanged.emit());
    }
    get subscriptions() {
        return this.member.subscriptions
            .map((s) => this.room._getSubscription(s.id))
            .filter((s) => s.stream);
    }
    _listenRoomEvent() {
        this.room.onPublicationUnsubscribed.add((e) => {
            if (e.subscription.subscriber._member.id ===
                this._local.id) {
                this.onPublicationUnsubscribed.emit(e);
            }
        });
    }
    /** internal */
    _publishAsP2P(stream, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const publication = yield this._local.publish(stream, options);
            const roomPublication = this.room._addPublication(publication);
            this.onStreamPublished.emit({ publication: roomPublication });
            return roomPublication;
        });
    }
    /** internal */
    _publishAsSFU(stream, options = {}) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (stream instanceof core_1.LocalDataStream) {
                throw errors_1.errors.sfuPublicationNotSupportDataStream;
            }
            options.type = 'sfu';
            options.maxSubscribers = (_a = options.maxSubscribers) !== null && _a !== void 0 ? _a : const_1.defaultMaxSubscribers;
            const origin = yield this._local.publish(stream, options);
            const bot = this.room._channel.members.find((m) => m.subtype === sfu_bot_1.SFUBotMember.subtype);
            if (!bot) {
                const authToken = this._context.authToken;
                if (authToken.isSfuCreateBotEnabled({
                    id: this.room._channel.id,
                    name: this.room._channel.name,
                })) {
                    throw sfu_bot_1.errors.sfuBotNotInChannel;
                }
                else {
                    throw sfu_bot_1.errors.insufficientPermissions;
                }
            }
            const forwarding = yield bot.startForwarding(origin, {
                maxSubscribers: options.maxSubscribers,
            });
            const relayingPublication = forwarding.relayingPublication;
            const roomPublication = this.room._addPublication(relayingPublication);
            this.onStreamPublished.emit({ publication: roomPublication });
            return roomPublication;
        });
    }
    /** internal */
    _unpublishAsP2P(target) {
        return __awaiter(this, void 0, void 0, function* () {
            const publicationId = typeof target === 'string' ? target : target.id;
            this._local.unpublish(publicationId).catch((error) => {
                log.error('_unpublishAsP2P error', error, { target }, this.toJSON());
            });
            const { publication } = yield this.room.onStreamUnpublished
                .watch((e) => e.publication.id === publicationId, this._context.config.rtcApi.timeout)
                .catch((error) => {
                throw [Object.assign(Object.assign({}, errors_1.errors.timeout), { detail: 'onStreamUnpublished' }), error];
            });
            this.onStreamUnpublished.emit({ publication });
        });
    }
    /** internal */
    _unpublishAsSFU(target) {
        return __awaiter(this, void 0, void 0, function* () {
            const publicationId = typeof target === 'string' ? target : target.id;
            const publication = this.room._getPublication(publicationId);
            const origin = publication._publication.origin;
            if (!origin) {
                throw [errors_1.errors.publicationNotHasOrigin];
            }
            this._local.unpublish(origin.id).catch((error) => {
                log.error('_unpublishAsSFU error', error, { target }, this.toJSON());
            });
            yield this.room.onStreamUnpublished
                .watch((e) => e.publication.id === publicationId, this._context.config.rtcApi.timeout)
                .catch((error) => {
                throw [Object.assign(Object.assign({}, errors_1.errors.timeout), { detail: 'onStreamUnpublished' }), error];
            });
            this.onStreamUnpublished.emit({ publication });
        });
    }
    subscribe(target, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const publicationId = typeof target === 'string' ? target : target.id;
            const { subscription, stream } = yield this._local.subscribe(publicationId, options);
            const roomSubscription = this.room._addSubscription(subscription);
            return {
                subscription: roomSubscription,
                stream: stream,
            };
        });
    }
    unsubscribe(target) {
        return __awaiter(this, void 0, void 0, function* () {
            const subscriptionId = typeof target === 'string' ? target : target.id;
            this._local.unsubscribe(subscriptionId).catch((error) => {
                log.error('unsubscribe error', error, { target }, this.toJSON());
            });
            yield this.room.onPublicationUnsubscribed
                .watch((e) => e.subscription.id === subscriptionId, this._context.config.rtcApi.timeout)
                .catch((error) => {
                throw [
                    Object.assign(Object.assign({}, errors_1.errors.timeout), { detail: 'onPublicationUnsubscribed' }),
                    error,
                ];
            });
        });
    }
}
exports.LocalRoomMemberBase = LocalRoomMemberBase;
