// @skyway-sdk/room/dist/member/local/sfu.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalSFURoomMemberImpl = void 0;
const common_1 = require("@skyway-sdk/common");
const errors_1 = require("../../errors");
const util_1 = require("../../util");
const base_1 = require("./base");
const log = new common_1.Logger('packages/room/src/member/local/sfu.ts');
/**@internal */
class LocalSFURoomMemberImpl extends base_1.LocalRoomMemberBase {
    /**@private */
    // biome-ignore lint/complexity/noUselessConstructor: Private constructor is intentional to control instantiation
    constructor(member, room) {
        super(member, room);
    }
    publish(stream, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            if (options.type && options.type !== 'sfu') {
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.publish',
                    context: this._context,
                    room: this.room,
                    info: errors_1.errors.invalidPublicationTypeForSFURoom,
                    path: log.prefix,
                });
            }
            const roomPublication = yield this._publishAsSFU(stream, options).catch((errorInfo) => {
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.publish',
                    context: this._context,
                    room: this.room,
                    info: errorInfo,
                    path: log.prefix,
                });
            });
            return roomPublication;
        });
    }
    /**
     * @description [japanese] Room上のStreamをUnPublishする
     */
    unpublish(target) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._unpublishAsSFU(target).catch((e) => {
                const [errorInfo, error] = e;
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.unpublish',
                    context: this._context,
                    room: this.room,
                    info: errorInfo,
                    path: log.prefix,
                    error,
                });
            });
        });
    }
    /**
     * @description [japanese] MemberがSubscribeしているStreamのSubscriptionをUnSubscribeする
     */
    unsubscribe(target) {
        const _super = Object.create(null, {
            unsubscribe: { get: () => super.unsubscribe }
        });
        return __awaiter(this, void 0, void 0, function* () {
            yield _super.unsubscribe.call(this, target).catch((e) => {
                const [errorInfo, error] = e;
                throw (0, util_1.createError)({
                    operationName: 'LocalSFURoomMemberImpl.unsubscribe',
                    context: this._context,
                    room: this.room,
                    info: errorInfo,
                    path: log.prefix,
                    error,
                });
            });
        });
    }
}
exports.LocalSFURoomMemberImpl = LocalSFURoomMemberImpl;
